### PR TITLE
ODC-6778: fix devcatalog types api issue

### PR DESCRIPTION
**Pull request by debsmita1**
Related to https://github.com/openshift/api/pull/1256

### CODE REVIEW

**Review by jerolimov — APPROVED** [review]
With this change finally the e2e tests in https://github.com/openshift/console-operator/pull/676 passes. Looks like his pointer is required to make the `types` optional for existing resources.

Our manual tests all work fine, but it looks like some client API validates the data and missed the state if `types` isn't configured.

@JoelSpeed @deads2k can we get this in also if it's not common?

/lgtm
/cc @JoelSpeed @deads2k

File: operator/v1/types_console.go [hunk 1]
```diff
@@ -194,7 +194,7 @@ type DeveloperConsoleCatalogCustomization struct {
 	// types allows enabling or disabling of sub-catalog types that user can see in the Developer catalog.
 	// When omitted, all the sub-catalog types will be shown.
 	// +optional
-	Types DeveloperConsoleCatalogTypesState `json:"types"`
+	Types *DeveloperConsoleCatalogTypes `json:"types,omitempty"`
```
Comment (JoelSpeed) [inline operator/v1/types_console.go:197]: Did you try with just omitempty and without the pointer? We have used this successfully in other areas of the API codebase. The pointer just adds complexity

Comment (jerolimov) [reply operator/v1/types_console.go:197]: @JoelSpeed Yes we tried first with `omitempty` only but had the same e2e CI issues. After that we added the pointer and now its working fine.

**Comment by jerolimov** [comment]
Hey @JoelSpeed @deads2k @sanketpathak,

@Debsmita and I have tested this API change together with https://github.com/openshift/console-operator/pull/676 on a cluster bot instance. Sorry I missed to record it. :man_shrugging: 

Manual testing was successful :+1: and also the e2e tests run fine now. :partying_face: 

Some details:

1. When defining just `types: {}` our UI says the state is missing (it's correct), but you can save it anyway (like the cli would do it) and the state was automatically set to `Enabled`.
2. We've verified that it's possible to save `Enabled`, `Disabled` catalog lists and that mixing `state: Disabled` and an `enabled` list, etc. doesn't work. All works fine. We've tested also the CEL rules on perspectives not on k8s 1.25 and it works as expected.
3. And finally, we checked that the console `ConfigMap` was updated and that the JS global variable (`window.SERVER_FLAGS.developerCatalogTypes`) in the UI was exported correctly.

**Review by jerolimov — APPROVED** [review]
/lgtm

**Comment by jerolimov** [comment]
As we have tested this on a cluster bot and in agreement with @sanketpathak we add QE-approved for this change:
/label qe-approved

Following our no-FF process propagate the PX and Docs ACK from our epic:
/label px-approved
/label docs-approved

**Comment by jerolimov** [comment]
/retitle ODC-6778: fix devcatalog types api issue

File: operator/v1/types_console.go [hunk 2]
```diff
@@ -156,14 +156,15 @@ const (
...
+// DeveloperConsoleCatalogTypes defines the state of the sub-catalog types.
+// +kubebuilder:validation:XValidation:rule="self.state == 'Enabled' ? true : !has(self.enabled)",message="enabled is forbidden when state is not Enabled"
+// +kubebuilder:validation:XValidation:rule="self.state == 'Disabled' ? true : !has(self.disabled)",message="disabled is forbidden when state is not Disabled"
 // +union
-type DeveloperConsoleCatalogTypesState struct {
+type DeveloperConsoleCatalogTypes struct {
 	// state defines if a list of catalog types should be enabled or disabled.
 	// +unionDiscriminator
 	// +kubebuilder:validation:Enum:="Enabled";"Disabled";
+	// +kubebuilder:default:="Enabled"
```
Comment (JoelSpeed) [inline operator/v1/types_console.go:167]: Something I forgot, can you add

// +default=Enabled


As well as this and regenerate the openapi definition, this is required for the generated apply client

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Pull request merged**